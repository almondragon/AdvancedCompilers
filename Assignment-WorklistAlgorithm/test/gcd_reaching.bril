# GCD: Greatest Common Divisor
# Euclidean algorithm

# input: two positive integer - op1, op2
# output: one positive integer - gcd(op1, op2)

@main (op1: int, op2: int) {
#ARGS: reaching
  # const
  vc0: int = const 0;
  # take two input ops, first iteration
  v0: int = id op1;
  v1: int = id op2;
.cmpval:
  v2: bool = lt v0 v1;
  br v2 .if1 .else1;
.if1:
  v3: int = sub v1 v0;
  jmp .loopbound;
.else1:
  v3: int = sub v0 v1;
  jmp .loopbound;
  # check results
.loopbound:
  v4: bool = eq v3 vc0;
  br v4 .programend .updateval;
.updateval:
  br v2 .if2 .else2;
  # update v1
.if2:
  v1: int = id v3;
  jmp .cmpval;
  # update v0
.else2:
  v0: int = id v3;
  jmp .cmpval;
  # print out the results
.programend:
  print v1;
}
