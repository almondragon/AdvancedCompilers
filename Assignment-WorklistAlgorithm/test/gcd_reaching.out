b1:
  in:  âˆ…
  out: ('v0', 'b1'), ('v1', 'b1'), ('vc0', 'b1')
cmpval:
  in:  ('v0', 'b1'), ('v0', 'else2'), ('v1', 'b1'), ('v1', 'if2'), ('v2', 'cmpval'), ('v3', 'else1'), ('v3', 'if1'), ('v4', 'loopbound'), ('vc0', 'b1')
  out: ('v0', 'b1'), ('v0', 'else2'), ('v1', 'b1'), ('v1', 'if2'), ('v2', 'cmpval'), ('v3', 'else1'), ('v3', 'if1'), ('v4', 'loopbound'), ('vc0', 'b1')
if1:
  in:  ('v0', 'b1'), ('v0', 'else2'), ('v1', 'b1'), ('v1', 'if2'), ('v2', 'cmpval'), ('v3', 'else1'), ('v3', 'if1'), ('v4', 'loopbound'), ('vc0', 'b1')
  out: ('v0', 'b1'), ('v0', 'else2'), ('v1', 'b1'), ('v1', 'if2'), ('v2', 'cmpval'), ('v3', 'if1'), ('v4', 'loopbound'), ('vc0', 'b1')
else1:
  in:  ('v0', 'b1'), ('v0', 'else2'), ('v1', 'b1'), ('v1', 'if2'), ('v2', 'cmpval'), ('v3', 'else1'), ('v3', 'if1'), ('v4', 'loopbound'), ('vc0', 'b1')
  out: ('v0', 'b1'), ('v0', 'else2'), ('v1', 'b1'), ('v1', 'if2'), ('v2', 'cmpval'), ('v3', 'else1'), ('v4', 'loopbound'), ('vc0', 'b1')
loopbound:
  in:  ('v0', 'b1'), ('v0', 'else2'), ('v1', 'b1'), ('v1', 'if2'), ('v2', 'cmpval'), ('v3', 'else1'), ('v3', 'if1'), ('v4', 'loopbound'), ('vc0', 'b1')
  out: ('v0', 'b1'), ('v0', 'else2'), ('v1', 'b1'), ('v1', 'if2'), ('v2', 'cmpval'), ('v3', 'else1'), ('v3', 'if1'), ('v4', 'loopbound'), ('vc0', 'b1')
updateval:
  in:  ('v0', 'b1'), ('v0', 'else2'), ('v1', 'b1'), ('v1', 'if2'), ('v2', 'cmpval'), ('v3', 'else1'), ('v3', 'if1'), ('v4', 'loopbound'), ('vc0', 'b1')
  out: ('v0', 'b1'), ('v0', 'else2'), ('v1', 'b1'), ('v1', 'if2'), ('v2', 'cmpval'), ('v3', 'else1'), ('v3', 'if1'), ('v4', 'loopbound'), ('vc0', 'b1')
if2:
  in:  ('v0', 'b1'), ('v0', 'else2'), ('v1', 'b1'), ('v1', 'if2'), ('v2', 'cmpval'), ('v3', 'else1'), ('v3', 'if1'), ('v4', 'loopbound'), ('vc0', 'b1')
  out: ('v0', 'b1'), ('v0', 'else2'), ('v1', 'if2'), ('v2', 'cmpval'), ('v3', 'else1'), ('v3', 'if1'), ('v4', 'loopbound'), ('vc0', 'b1')
else2:
  in:  ('v0', 'b1'), ('v0', 'else2'), ('v1', 'b1'), ('v1', 'if2'), ('v2', 'cmpval'), ('v3', 'else1'), ('v3', 'if1'), ('v4', 'loopbound'), ('vc0', 'b1')
  out: ('v0', 'else2'), ('v1', 'b1'), ('v1', 'if2'), ('v2', 'cmpval'), ('v3', 'else1'), ('v3', 'if1'), ('v4', 'loopbound'), ('vc0', 'b1')
programend:
  in:  ('v0', 'b1'), ('v0', 'else2'), ('v1', 'b1'), ('v1', 'if2'), ('v2', 'cmpval'), ('v3', 'else1'), ('v3', 'if1'), ('v4', 'loopbound'), ('vc0', 'b1')
  out: ('v0', 'b1'), ('v0', 'else2'), ('v1', 'b1'), ('v1', 'if2'), ('v2', 'cmpval'), ('v3', 'else1'), ('v3', 'if1'), ('v4', 'loopbound'), ('vc0', 'b1')
